tsdb/engine/tsm1/tsm1.go

```diff
@@ -537,7 +576,47 @@ func (e *Engine) Compact(fullCompaction bool) error {
 		}
 
 		// we've got the write lock and the files are all there
-		break
+		return
+	}
+}
+
+// compactToNewFiles will compact the passed in data files into as few files as possible
+func (e *Engine) compactToNewFiles(minTime, maxTime int64, files dataFiles) []*os.File {
+	fileName := e.nextFileName()
+	e.logger.Printf("Starting compaction in %s of %d files to new file %s", e.path, len(files), fileName)
+
+	compaction := newCompactionJob(files, minTime, maxTime, e.MaxFileSize, e.MaxPointsPerBlock)
+	compaction.newCurrentFile(fileName)
+
+	// loop writing data until we've read through all the files
+	for {
```

Comment:
It could, but only if the `compactionJob` has a reference to the engine because it needs to get the next file name, which requires access to the files lock.